sound/soc/soc-pcm.c

```diff
@@ -2473,7 +2473,8 @@ int dpcm_be_dai_prepare(struct snd_soc_pcm_runtime *fe, int stream)
 
 		if ((be->dpcm[stream].state != SND_SOC_DPCM_STATE_HW_PARAMS) &&
 		    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_STOP) &&
-		    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_SUSPEND))
+		    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_SUSPEND) &&
+		    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED))
```

Comment:
is this correct?
If you restart playing, the state should be RELEASE, and I am not sure it makes sense to transition from RELEASE to PREPARE?
Can you clarify what assumptions you made here and what state machine you based this patch on?

Reply:
@plbossart My comments is ambiguous. The release is not meaning RELEASE state. It means to release from pause, in another word, continue to playback.
Image the flow:
playback (1) -> pause (2) -> suspend (3) -> resume (4) -> release from pause (continue to playback) (5)
In step 5, we need reset the hw_params as S3 has happened. And as we are release from pause, the state is SND_SOC_DPCM_STATE_PAUSED. So we must let the asoc call prepare() callback to restore the hw_params, just like we are in SUPSEND status.

We are not changing the state to PREPARE here.

Reply:
I am still confused as to when dpcm_be_dai_prepare() will be called.
If you are in PAUSED state, under what circumstances will you reach this function?

Reply:
ALSA core will call dpcm_be_dai_prepare() when we continue to playback from pause. The state is PAUSED, and we are now releasing from pause. We are in step 5: release from pause (continue to playback)

Reply:
A little correction: "ALSA core will call dpcm_be_dai_prepare() when we continue to playback from pause" => "ALSA core will call dpcm_be_dai_prepare() when we continue to playback from pause after resume"

Reply:
Actually we are in "pause release" processing, and the state is in PAUSED. This is ASoC criterion. After prepare() callback is called, the state will be changed to PREPARE.

Reply:
@plbossart I think the problem is that only RUNNING or DRAINING pcm's are suspended when the system enters S3. If a stream is paused prior to suspend and the user releases the pause after resume, only the pause-release trigger gets called directly. There will be no prepare() ioctl as with running or draining streams

Reply:
@ranj063 but why should there be in the general case a prepare after a pause/release? this doesn't seem right to me. This is a platform-dependent limitation that you are pushing in the core.

Reply:
@plbossart pause/release will not call prepare(). prepare() is called because the system has resumed form suspend. Alsa will call it automatically.

> @ranj063 but why should there be in the general case a prepare after a pause/release? this doesn't seem right to me. This is a platform-dependent limitation that you are pushing in the core.

Reply:
@libinyang Pierre is right. We cannot have prepare() in the regular case. You are arguing from a SOF point of view. SOF doesnt do anything in prepare() if not resuming from suspend. But other drivers might and we cannot guarantee that it will break anything. 

So, for SOF, we will not have a prepare() for paused stream when resuming from S3. So our only option is to set the trigger order to POST and set hw_params in the PAUSE_RELEASE trigger if the internal flag is set.

Reply:
@plbossart @libinyang I did some experiments today and because we have removed the INFO_RESUME flag in SOF, the prepare() ioctl gets called for the FE dai even when it is paused and the stream is re-started. But because prepare() isnt called for the BE dai, link DMA fails to resume.

I think the fix in this PR is the only way to fix the problem.

Reply:
it's a bug if the stream is restarted after a resume without the user sending an explicit RELEASE command.

Reply:
@plbossart what I meant was that the stream is restarted after the user presses the pause release button. So after resume, we still get the START trigger and not the PAUSE_RELEASE trigger. But prepare() happens only for FE.

Reply:
@plbossart If there is no RELEASE, the stream will not be restarted after resume in pause state. And what's more, if a stream is restarted after resume, we definitely need call prepare() to setup hw_params.

So if there is a stream restart after resume without release cmd, I would like to add such state into this judgement.

> it's a bug if the stream is restarted after a resume without the user sending an explicit RELEASE command.

Reply:
@plbossart @libinyang @ranj063 .   I am also confused at the power management in ASoC, especially about flags like PAUSE, RELEASE.  Can we send it to alsa email list and involve more people to discuss this ?